ledger/participant-integration-api/src/main/scala/platform/store/backend/common/Field.scala

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.backend.common
+
+import java.lang
+
+import scala.reflect.ClassTag
+
+/** @tparam FROM is an arbitrary type from which we can extract the data of interest for the particular column
+  * @tparam TO is the intermediary type of the result of the extraction.
+  *            FROM => TO functionality is intended to be injected at Schema definition time.
+  *            TO is not nullable, should express a clean Scala type
+  * @tparam CONVERTED is the (possibly primitive) type needed by the JDBC API
+  *                   TO => CONVERTED is intended to be injected at PGField definition time.
+  *                   CONVERTED might be nullable, primitive, boxed-type, whatever the JDBC API requires
+  */
+abstract class Field[FROM, TO, CONVERTED](implicit
+    classTag: ClassTag[CONVERTED]
+) {
+  def extract: FROM => TO
+  def convert: TO => CONVERTED
+  def selectFieldExpression(inputFieldName: String): String = inputFieldName
+
+  final def toArray(input: Vector[FROM]): Array[CONVERTED] =
+    input.view
+      .map(extract andThen convert)
+      .toArray(classTag)
+}
+
+abstract class TrivialField[FROM, TO](implicit classTag: ClassTag[TO]) extends Field[FROM, TO, TO] {
+  override def convert: TO => TO = identity
+}
+
+trait TrivialOptionalField[FROM, TO >: Null <: AnyRef] extends Field[FROM, Option[TO], TO] {
+  override def convert: Option[TO] => TO = _.orNull
+}
+
+case class StringField[FROM](extract: FROM => String) extends TrivialField[FROM, String]
+
+case class StringOptional[FROM](extract: FROM => Option[String])
+    extends TrivialOptionalField[FROM, String]
+
+case class Bytea[FROM](extract: FROM => Array[Byte]) extends TrivialField[FROM, Array[Byte]]
+
+case class ByteaOptional[FROM](extract: FROM => Option[Array[Byte]])
+    extends TrivialOptionalField[FROM, Array[Byte]]
+
+case class IntOptional[FROM](extract: FROM => Option[Int])
+    extends Field[FROM, Option[Int], java.lang.Integer] {
+  override def convert: Option[Int] => Integer = _.map(Int.box).orNull
+}
+
+case class Bigint[FROM](extract: FROM => Long) extends TrivialField[FROM, Long]
+
+case class SmallintOptional[FROM](extract: FROM => Option[Int])
+    extends Field[FROM, Option[Int], java.lang.Integer] {
+  override def selectFieldExpression(inputFieldName: String): String =
+    s"$inputFieldName::smallint"
```

Comment:
The double colon operator is not standard (not that "standard" would mean much when it comes to SQL dialects). I assume this is again a case where we refactor as necessary if we add more database backends?

Reply:
No!! This was completely not intentional, thanks for spotting that!
Unfortunately H2 supports the :: style cast (which I could not find evidence of in the docs), so even on my branch where H2 tests are green I could not have spotted that.
Changed it, and created/wired again the PGSmallintOptional for postgres

Reply:
the sql standard syntax is ```cast ( $inputFieldName as smallint)``` and that should work for oracle and (presumably) h2 so would be the better cross-platform way.  i believe the double colon is just postgres shorthand for the same

Reply:
correct, the double colon is for postgres for sure but also for others (hence my surprise seeing that work with H2)
but the main thing here to keep it simple: in this case for H2 we do not even need the cast, we need it only for postgres, so a specialization in the postgres strategy is called for IMHO